### PR TITLE
[note]: Migrated into new api.

### PR DESCRIPTION
**Pull request by rgo-odoo**
Task: https://www.odoo.com/web#id=13294&view_type=form&model=project.task&menu_id=3942&action=327
Pad: https://pad.odoo.com/p/r.ZG9OnTSXxZbnLSYM

### CODE REVIEW

File: addons/note/models/note_note.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+    name = fields.Text(compute='_compute_name', string='Note Summary', store=True)
+    user_id = fields.Many2one('res.users', string='Owner', default=lambda self: self.env.uid)
+    memo = fields.Html(string='Note Content')
+    sequence = fields.Integer()
+    stage_id = fields.Many2one('note.stage', compute='_compute_stage_id',
+       inverse='_inverse_stage_id', default=_get_default_stage_id, string='Stage')
+    stage_ids = fields.Many2many('note.stage', 'note_stage_rel',
+       'note_id', 'stage_id', string='Stages of Users')
+    open = fields.Boolean(string='Active', track_visibility='onchange', default=True)
```
Comment (WHPODOO) [inline addons/note/models/note_note.py]: open is preserve key word so should be change

File: addons/note/models/note_note.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    #For kanban view, get stage id
+    @api.model
+    def create(self, vals):
+        if not vals.get('stage_id'):
+            vals['stage_id'] = self.env.context.get('default_stage_id')
+        return super(NoteNote, self).create(vals)
+
+    #read the first line (convert html into text)
+    @api.one
```
Comment (WHPODOO) [inline addons/note/models/note_note.py]: Don't use @api.one

File: addons/note/models/note_note.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            for stage in note.stage_ids.filtered(lambda stage: stage.user_id == self.env.user):
+                note.stage_id = stage
+
+    def _inverse_stage_id(self):
+        for note in self:
+            if note.stage_id:
+                note.stage_ids = note.stage_id + \
+                note.stage_ids.filtered(lambda stage: stage.user_id != self.env.user)
+
+    @api.one
```
Comment (WHPODOO) [inline addons/note/models/note_note.py]: Don't use @api.one

File: addons/note/models/note_note.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    def _action_note_is_done(self):
+        self.open = False
+        self.date_done = fields.Date.today()
+
+    @api.one
```
Comment (WHPODOO) [inline addons/note/models/note_note.py]: Don't use @api.one

File: addons/note/tests/test_note.py [hunk 5]
```diff
@@ -1,25 +1,21 @@
...
-            'groups_id': [(4, group_id)],
-        })
+        demo_user.write({'groups_id': [(4, group.id)]})
 
         # must not fail
-        Users.create(cr, demo_user, {
+        self.env(user=demo_user.id)['res.users'].create({
             'name': 'test bug lp:1156215',
             'login': 'lp_1156215',
         })
```
Comment (WHPODOO) [inline addons/note/tests/test_note.py]: After this creation it should be check that its created or not, Its just suggetion. :sake:

File: addons/note/models/note_note.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    @api.model
+    def create(self, vals):
+        if not vals.get('stage_id'):
+            vals['stage_id'] = self.env.context.get('default_stage_id')
+        return super(NoteNote, self).create(vals)
+
+    #read the first line (convert html into text)
+    @api.depends('memo')
+    def _compute_name(self):
+        self.name = (self.memo and html2plaintext(self.memo) or "").\
```
Comment (atp-odoo) [inline addons/note/models/note_note.py]: iteration required when used @api.depends or @api.multi

File: addons/note/models/note_note.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    def _inverse_stage_id(self):
+        for note in self:
+            if note.stage_id:
+                note.stage_ids = note.stage_id + \
+                note.stage_ids.filtered(lambda stage: stage.user_id != self.env.user)
+
+    @api.multi
+    def onclick_note_is_done(self):
+        self.ensure_one()
```
Comment (atp-odoo) [inline addons/note/models/note_note.py]: remove ensure_one().

File: addons/note/models/note_note.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+    name = fields.Text(compute='_compute_name', string='Note Summary', store=True)
+    user_id = fields.Many2one('res.users', string='Owner', default=lambda self: self.env.uid)
+    memo = fields.Html(string='Note Content')
+    sequence = fields.Integer()
+    stage_id = fields.Many2one('note.stage', compute='_compute_stage_id',
+       inverse='_inverse_stage_id', default=_get_default_stage_id, string='Stage')
+    stage_ids = fields.Many2many('note.stage', 'note_stage_rel',
+       'note_id', 'stage_id', string='Stages of Users')
+    open_active = fields.Boolean(string='Active', track_visibility='onchange', default=True)
```
Comment (atp-odoo) [inline addons/note/models/note_note.py]: use active name and when you changed field name use oldname in field.

File: addons/note/models/note_note.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        for note in self:
+            if note.stage_id:
+                note.stage_ids = note.stage_id + \
+                note.stage_ids.filtered(lambda stage: stage.user_id != self.env.user)
+
+    @api.multi
+    def onclick_note_is_done(self):
+        self.ensure_one()
+        self.open_active = False
+        self.date_done = fields.Date.today()
```
Comment (atp-odoo) [inline addons/note/models/note_note.py]: use directly self.write({'open_active': False, 'date_done': fields.date.today()})

File: addons/note/models/note_note.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+    @api.multi
+    def onclick_note_not_done(self):
+        self.ensure_one()
+        self.open_active = True
```
Comment (atp-odoo) [inline addons/note/models/note_note.py]: same as above remove ensure_one() and use write method

File: addons/note/models/note_note.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+                else:
+                    result = []
+            return result
+        else:
+            return super(NoteNote, self).read_group(domain, fields, groupby,
+                offset=offset, limit=limit, orderby=orderby, lazy=lazy)
+
+    def _notification_get_recipient_groups(self, message, recipients):
+        res = super(NoteNote, self)._notification_get_recipient_groups(message, recipients)
+        new_action_id = self.env.ref('note.action_note_note').id
```
Comment (atp-odoo) [inline addons/note/models/note_note.py]: if we need id only then better to use xml_id_res_id method of ir.model.data

**Comment by atp-odoo** [comment]
use from odoo import api, fields, models

File: addons/note/models/note_note.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.tools import html2plaintext
+
+
+class NoteNote(models.Model):
```
Comment (atp-odoo) [inline addons/note/models/note_note.py]: class name should be "Note" instead of "NoteNote"

File: addons/note/tests/test_note.py [hunk 13]
```diff
@@ -1,25 +1,23 @@
...
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.tests import common
+from odoo.tests.common import TransactionCase
 
-class TestNote(common.TransactionCase):
+
+class TestNote(TransactionCase):
 
     def test_bug_lp_1156215(self):
```
Comment (atp-odoo) [inline addons/note/tests/test_note.py]: change method name give proper name..

File: addons/note/tests/test_note.py [hunk 14]
```diff
@@ -1,25 +1,23 @@
...
         """ensure any users can create new users"""
-        cr, uid = self.cr, self.uid
-        IMD = self.registry('ir.model.data')
-        Users = self.registry('res.users')
 
-        _, demo_user = IMD.get_object_reference(cr, uid, 'base', 'user_demo')
-        _, group_id = IMD.get_object_reference(cr, uid, 'base', 'group_erp_manager')
+        demo_user = self.env.ref('base.user_demo')
+        group = self.env.ref('base.group_erp_manager')
```
Comment (atp-odoo) [inline addons/note/tests/test_note.py]: if want id then no need to use self.env.ref you can directly get it using self.ref('..')

File: addons/note/views/note_note_views.xml [hunk 15]
```diff
@@ -1,43 +1,7 @@
...
-        <field name="res_model">note.stage</field>
-        <field name="view_mode">tree,form</field>
-        <field name="domain">[('user_id','=',uid)]</field>
-    </record>
-
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
     <!-- New note Kanban View -->
-    <record model="ir.ui.view" id="view_note_note_kanban">
+    <record id="note_note_view_kanban" model="ir.ui.view">
```
Comment (atp-odoo) [inline addons/note/views/note_note_views.xml]: why did you change id ?

File: addons/note/models/note.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    #For kanban view, get stage id
+    @api.model
+    def create(self, vals):
+        if not vals.get('stage_id'):
+            vals['stage_id'] = self.env.context.get('default_stage_id')
+        return super(Note, self).create(vals)
+
+    #read the first line (convert html into text)
+    @api.depends('memo')
+    def _compute_name(self):
```
Comment (atp-odoo) [inline addons/note/models/note.py]: define this method before create

File: addons/note/models/note.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+                    result = [{  # notes for unknown stage
+                        '__context': {'group_by': groupby[1:]},
+                        '__domain': domain,
+                        'stage_id': False,
+                        'stage_id_count': nb_notes_ws
+                    }]
+                else:
+                    result = []
+            return result
+        else:
```
Comment (atp-odoo) [inline addons/note/models/note.py]: no need of else here ....

File: addons/note/models/note.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+            return super(Note, self).read_group(domain, fields, groupby,
+                offset=offset, limit=limit, orderby=orderby, lazy=lazy)
+
+    def _notification_get_recipient_groups(self, message, recipients):
```
Comment (atp-odoo) [inline addons/note/models/note.py]: are you sure no need of decorator ?

File: addons/note/models/note_stage.py [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class NoteStage(models.Model):
+    """ Category of Note """
+    _name = "note.stage"
+    _description = "Note Stage"
+    _order = 'sequence asc'
```
Comment (atp-odoo) [inline addons/note/models/note_stage.py]: default is ascending order . _order = 'sequence'

File: addons/note/models/note_stage.py [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+    name = fields.Char(string='Stage Name', translate=True, required=True)
+    sequence = fields.Integer(help="Used to order the note stages", default=1)
+    user_id = fields.Many2one('res.users', string='Owner', help="Owner of the note stage.",
+                  required=True, ondelete='cascade', default=lambda self: self.env.uid)
+    fold = fields.Boolean(string='Folded by Default', default=0)
```
Comment (atp-odoo) [inline addons/note/models/note_stage.py]: default value for boolean field is False. remove default=0

File: addons/note/models/res_users.py [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+    _inherit = 'res.users'
+
+    @api.model
+    def create(self, data):
+        user = super(ResUsers, self).create(data)
+        if self.has_group('base.group_user'):
+            for n in range(5):
+                xmlid = 'note_stage_%02d' % (n,)
+                stage = self.sudo().env.ref('note.' + xmlid, raise_if_not_found=False)
+                if stage:
```
Comment (atp-odoo) [inline addons/note/models/res_users.py]: no need of if statement..

File: addons/note/tests/test_note.py [hunk 22]
```diff
@@ -1,25 +1,22 @@
...
+        demo_user = self.env.ref('base.user_demo')
 
-        Users.write(cr, uid, [demo_user], {
-            'groups_id': [(4, group_id)],
-        })
+        demo_user.write({'groups_id': [(4, self.ref('base.group_erp_manager'))]})
 
         # must not fail
-        Users.create(cr, demo_user, {
+        UserLogin = self.env(user=demo_user)['res.users'].create({
```
Comment (atp-odoo) [inline addons/note/tests/test_note.py]: self.env['res.users'].sudo(demo_user).create(

File: addons/note/models/note.py [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+       'note_id', 'stage_id', string='Stages of Users')
+    open_active = fields.Boolean(string='Active', track_visibility='onchange', default=True, oldname='open')
+    date_done = fields.Date()
+    color = fields.Integer(string='Color Index')
+    tag_ids = fields.Many2many('note.tag', 'note_tags_rel', 'note_id', 'tag_id', string='Tags')
+
+    #read the first line (convert html into text)
+    @api.depends('memo')
+    def _compute_name(self):
+        for data in self:
```
Comment (atp-odoo) [inline addons/note/models/note.py]: for data in self ==> for note in self

File: addons/note/models/note.py [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+        return super(Note, self).create(vals)
+
+    @api.multi
+    def _compute_stage_id(self):
+        for note in self:
+            for stage in note.stage_ids.filtered(lambda stage: stage.user_id == self.env.user):
+                note.stage_id = stage
+
+    def _inverse_stage_id(self):
+        for note in self:
```
Comment (atp-odoo) [inline addons/note/models/note.py]: self.filtered('stage_id') then no need of if condition.

File: addons/note/models/note.py [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+            if note.stage_id:
+                note.stage_ids = note.stage_id + \
+                note.stage_ids.filtered(lambda stage: stage.user_id != self.env.user)
+
+    @api.multi
+    def onclick_note_is_done(self):
```
Comment (atp-odoo) [inline addons/note/models/note.py]: change name: 
onclick_note_is_done ==> action_done

File: addons/note/models/note.py [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+        self.write({'open_active': False, 'date_done': fields.date.today()})
+
+    @api.multi
+    def onclick_note_not_done(self):
```
Comment (atp-odoo) [inline addons/note/models/note.py]: nclick_note_not_done ==> action_active

File: addons/note/models/note_stage.py [hunk 27]
```diff
@@ -0,0 +1,17 @@
...
+    """ Category of Note """
+    _name = "note.stage"
+    _description = "Note Stage"
+    _order = 'sequence'
+
+    name = fields.Char(string='Stage Name', translate=True, required=True)
+    sequence = fields.Integer(help="Used to order the note stages", default=1)
+    user_id = fields.Many2one('res.users', string='Owner', help="Owner of the note stage.",
+                  required=True, ondelete='cascade', default=lambda self: self.env.uid)
+    fold = fields.Boolean(string='Folded by Default', default=False)
```
Comment (atp-odoo) [inline addons/note/models/note_stage.py]: bydefault boolean value is not False ?

File: addons/note/models/note.py [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+
+    name = fields.Text(compute='_compute_name', string='Note Summary', store=True)
+    user_id = fields.Many2one('res.users', string='Owner', default=lambda self: self.env.uid)
+    memo = fields.Html(string='Note Content')
+    sequence = fields.Integer()
+    stage_id = fields.Many2one('note.stage', compute='_compute_stage_id',
+       inverse='_inverse_stage_id', default=_get_default_stage_id, string='Stage')
+    stage_ids = fields.Many2many('note.stage', 'note_stage_rel',
+       'note_id', 'stage_id', string='Stages of Users')
+    open_active = fields.Boolean(string='Active', track_visibility='onchange', default=True, oldname='open')
```
Comment (atp-odoo) [inline addons/note/models/note.py]: keep separate commit for this field name change...

File: addons/note/models/note.py [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+                if nb_notes_ws:
+                    result = [{  # notes for unknown stage
+                        '__context': {'group_by': groupby[1:]},
+                        '__domain': domain,
+                        'stage_id': False,
+                        'stage_id_count': nb_notes_ws
+                    }]
+                else:
+                    result = []
+            return result
```
Comment (atp-odoo) [inline addons/note/models/note.py]: where is super call for read_group  ? what if no groupby[0] == "stage_id" ?
keep this back
 return super(Note, self).read_group(domain, fields, groupby,
                offset=offset, limit=limit, orderby=orderby, lazy=lazy)

File: addons/note/models/res_users.py [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+    def create(self, data):
+        user = super(ResUsers, self).create(data)
+        NoteStage = self.env['note.stage']
+        if self.has_group('base.group_user'):
+            for n in range(5):
+                xmlid = 'note_stage_%02d' % (n,)
+                try:
+                    stage_id = self.sudo().env.ref('note.' + xmlid).id
+                except ValueError:
+                    continue
```
Comment (atp-odoo) [inline addons/note/models/res_users.py]: this is not required. just pass this parametere "raise_if_not_found=False" to self.env.ref() function.

File: addons/note/models/note.py (hunk 28 above)
Comment (WHPODOO) [inline addons/note/models/note.py]: @atp-odoo  :+1:

File: addons/note/models/note.py [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+
+    def _inverse_stage_id(self):
+        for note in self.filtered('stage_id'):
+            note.stage_ids = note.stage_id + \
+            note.stage_ids.filtered(lambda stage: stage.user_id != self.env.user)
+
+    #For kanban view, get stage id
+    @api.model
+    def create(self, vals):
+        if not vals.get('stage_id'):
```
Comment (WHPODOO) [inline addons/note/models/note.py]: You can use also use this way : 
```
if not vals['stage_id']:
```

File: addons/note/models/note.py [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+
+    @api.multi
+    def action_active(self):
+        self.write({'open_active': True})
+
+    #writing method (no modification of values)
+    @api.model
+    def name_create(self, name):
+        note = self.create({'memo': name})
+        return note.name_get()[0]
```
Comment (WHPODOO) [inline addons/note/models/note.py]: Save some bytes one time use variable : 
```
@api.model
def name_create(self, name):
    return self.create({'memo': name}).name_get()[0]

```

File: addons/note/models/note.py [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+                result = [{  # notes by stage for stages user
+                    '__context': {'group_by': groupby[1:]},
+                    '__domain': domain + [('stage_ids.id', '=', stage.id)],
+                    'stage_id': (stage.id, stage.name),
+                    'stage_id_count': self.search_count(domain + [('stage_ids', '=', stage.id)]),
+                    '__fold': stage.fold,
+                } for stage in stages]
+
+                #note without user's stage
+                nb_notes_ws = self.search_count(domain + [('stage_ids', 'not in', stages.ids)])
```
Comment (WHPODOO) [inline addons/note/models/note.py]: you can use ***read_group*** instead of ***search_count***

File: addons/note/models/note.py [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+                        'stage_id_count': nb_notes_ws
+                    }]
+                else:
+                    result = []
+            return result
+
+    @api.multi
+    def _notification_get_recipient_groups(self, message, recipients):
+        res = super(Note, self)._notification_get_recipient_groups(message, recipients)
+        new_action_id = self.env['ir.model.data'].xmlid_to_res_id('note.action_note_note')
```
Comment (WHPODOO) [inline addons/note/models/note.py]: you can use ***self.env.ref***

File: addons/note/views/note_views.xml [hunk 35]
```diff
@@ -166,7 +130,5 @@
 
     <menuitem name="Notes" id="menu_note_notes" sequence="3" action="note.action_note_note" web_icon="note,static/description/icon.png"/>
     <menuitem name="Configuration" id="menu_note_configuration" sequence="100" parent="menu_note_notes" groups="base.group_no_one"/>
-    <menuitem name="Stages" id="menu_notes_stage" parent="menu_note_configuration" action="note.action_note_stage" sequence="21" />
```
Comment (atp-odoo) [inline addons/note/views/note_views.xml]: why this menu removed? still exists in master branch..

File: addons/note/models/res_users.py [hunk 36]
```diff
@@ -0,0 +1,18 @@
...
+class ResUsers(models.Model):
+    _inherit = 'res.users'
+
+    @api.model
+    def create(self, data):
+        user = super(ResUsers, self).create(data)
+        NoteStage = self.env['note.stage']
+        if self.has_group('base.group_user'):
+            for n in range(5):
+                stage_id = self.sudo().env.ref('note.' + 'note_stage_%02d' % (n,), raise_if_not_found=False).id
```
Comment (atp-odoo) [inline addons/note/models/res_users.py]: instead of this : 'note.' + 'note_stage_%02d' % (n,) ==> do this way 
'note.note_stage_%02d' % (n,)

File: addons/note/models/res_users.py [hunk 37]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class ResUsers(models.Model):
+    _inherit = 'res.users'
+
+    @api.model
+    def create(self, data):
+        user = super(ResUsers, self).create(data)
+        NoteStage = self.env['note.stage']
+        if self.has_group('base.group_user'):
```
Comment (hmo-odoo) [inline addons/note/models/res_users.py]: should be check has group of newly created user not a current user

File: addons/note/models/res_users.py [hunk 38]
```diff
@@ -0,0 +1,18 @@
...
+            for n in range(5):
+                stage_id = self.sudo().env.ref('note.note_stage_%02d' % (n,), raise_if_not_found=False).id
```
Comment (hmo-odoo) [inline addons/note/models/res_users.py]: no need to access env.ref with sudo

File: addons/note/views/note_stage_views.xml [hunk 39]
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <!-- note Stage Form View -->
+    <record id="view_note_stage_form" model="ir.ui.view">
+      <field name="name">note.stage.form</field>
```
Comment (hmo-odoo) [inline addons/note/views/note_stage_views.xml]: give 4 white space . plz take care on indentation to look good format.

File: addons/note/models/note.py (hunk 28 above)
Comment (hmo-odoo) [inline addons/note/models/note.py]: what is meaning of open_active ? can not take only active ? field name should be either is_open or only active.

File: addons/note/models/res_users.py [hunk 40]
```diff
@@ -0,0 +1,17 @@
...
+class ResUsers(models.Model):
+    _inherit = 'res.users'
+
+    @api.model
+    def create(self, data):
+        user = super(ResUsers, self).create(data)
+        if user.has_group('base.group_user'):
+            for n in range(5):
+                stage_id = self.env.ref('note.note_stage_%02d' % (n,), raise_if_not_found=False)
+                stage_id.sudo().copy(default={'user_id': user.id})
```
Comment (rgo-odoo) [inline addons/note/models/res_users.py]: stage_id may contain 'False', hence, you need to check for it's value before copying it.

**Comment by hmo-odoo** [comment]
merged at https://github.com/odoo/odoo/commit/b3b0b23fe2094812eb9d609b43c04a4a70c60cbf
